File: android/app/src/main/java/com/pablomartin/encriptador/CryptoService.kt
package com.pablomartin.encriptador

import java.io.EOFException
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.GeneralSecurityException
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/** Contraseña incorrecta o archivo .enc dañado/manipulado (falló la verificación GCM). */
class ArchivoDanadoException(mensaje: String = Localization.t("crypto.archivoDanado")) :
    GeneralSecurityException(mensaje)

/** El .enc es de carpeta pero se pidió tratarlo como archivo individual, o viceversa. */
class TipoIncorrectoException(mensaje: String) : IllegalStateException(mensaje)

data class EntradaManifiesto(val relativa: String, val longitud: Long, val offset: Long)

/**
 * Cifra/descifra con AES-256-GCM por bloques de 64 KB (mismo esquema que la versión
 * de escritorio: cada bloque lleva su propio nonce y tag de autenticación, así que
 * una contraseña incorrecta o un archivo manipulado se detectan de inmediato). El
 * formato de archivo es compatible byte a byte con la versión Windows/Linux, aunque
 * el código de esta app es una implementación independiente, sin compartir nada.
 *
 * Un archivo individual: [salt 16][tipo=1][longitud int64][datos cifrados por bloques]
 * Una carpeta: [salt 16][tipo=2][longitud manifiesto int64][manifiesto cifrado][datos de cada archivo, uno atrás del otro]
 */
object CryptoService {
    private const val SALT_SIZE = 16
    private const val KEY_SIZE_BYTES = 32
    private const val PBKDF2_ITERATIONS = 600_000
    private const val TIPO_ARCHIVO = 1
    private const val TIPO_CARPETA = 2
    private const val MAX_ENTRADAS = 5_000_000
    private const val MAX_LONGITUD_RUTA = 32_000
    private const val MAX_LONGITUD_MANIFIESTO = 200_000_000L
    private const val TAMANO_CHUNK = 64 * 1024
    private const val NONCE_SIZE = 12
    private const val TAG_SIZE = 16

    const val EXTENSION = ".enc"

    fun esArchivoEncriptado(nombre: String) = nombre.endsWith(EXTENSION, ignoreCase = true)

    // ===================== Archivo individual =====================

    fun encriptarArchivo(entrada: InputStream, longitud: Long, salida: OutputStream, password: String) {
        val salt = ByteArray(SALT_SIZE).also { SecureRandom().nextBytes(it) }
        val clave = derivarClave(password, salt)
        salida.write(salt)
        salida.write(TIPO_ARCHIVO)
        escribirLong(salida, longitud)
        cifrarStream(entrada, salida, clave, longitud)
    }

    /** Lee el tipo de un .enc (archivo o carpeta) sin necesitar la contraseña: va sin cifrar. */
    fun esContenedorDeCarpeta(entrada: InputStream): Boolean {
        val salt = ByteArray(SALT_SIZE)
        leerCompleto(entrada, salt)
        return entrada.read() == TIPO_CARPETA
    }

    fun desencriptarArchivo(entrada: InputStream, salida: OutputStream, password: String) {
        val salt = ByteArray(SALT_SIZE)
        leerCompleto(entrada, salt)
        val tipo = entrada.read()
        if (tipo != TIPO_ARCHIVO) throw TipoIncorrectoException(Localization.t("crypto.tipoEsCarpeta"))

        val clave = derivarClave(password, salt)
        val longitud = leerLong(entrada)
        if (longitud < 0) throw ArchivoDanadoException()
        descifrarStream(entrada, salida, clave, longitud)
    }

    // ===================== Carpeta =====================

    /** Cada entrada de origen: ruta relativa (con '/') + su tamaño + una forma de abrirla para leer. */
    class FuenteArchivo(val relativa: String, val longitud: Long, val abrir: () -> InputStream)

    fun encriptarCarpeta(archivos: List<FuenteArchivo>, salida: OutputStream, password: String, progreso: ((Int, Int) -> Unit)? = null): Int {
        val salt = ByteArray(SALT_SIZE).also { SecureRandom().nextBytes(it) }
        val clave = derivarClave(password, salt)

        var offsetAcumulado = 0L
        val entradas = archivos.map { f ->
            val e = EntradaManifiesto(f.relativa, f.longitud, offsetAcumulado)
            offsetAcumulado += longitudCifrada(f.longitud)
            e
        }

        salida.write(salt)
        salida.write(TIPO_CARPETA)

        val manifiestoPlano = serializarManifiesto(entradas)
        escribirLong(salida, manifiestoPlano.size.toLong())
        cifrarStream(manifiestoPlano.inputStream(), salida, clave, manifiestoPlano.size.toLong())

        archivos.forEachIndexed { i, f ->
            f.abrir().use { cifrarStream(it, salida, clave, f.longitud) }
            progreso?.invoke(i + 1, archivos.size)
        }

        return archivos.size
    }

    class SesionCarpeta internal constructor(
        val entradas: List<EntradaManifiesto>,
        internal val clave: ByteArray,
    )

    /** Valida la contraseña y descifra solo el manifiesto (nombres/tamaños), sin tocar contenido de archivos. */
    fun abrirCarpeta(entrada: InputStream, password: String): SesionCarpeta {
        val salt = ByteArray(SALT_SIZE)
        leerCompleto(entrada, salt)
        val tipo = entrada.read()
        if (tipo != TIPO_CARPETA) throw TipoIncorrectoException(Localization.t("crypto.tipoEsArchivo"))

        val clave = derivarClave(password, salt)
        val manifiestoLen = leerLong(entrada)
        if (manifiestoLen < 0 || manifiestoLen > MAX_LONGITUD_MANIFIESTO) throw ArchivoDanadoException()

        val bufferSalida = java.io.ByteArrayOutputStream()
        descifrarStream(entrada, bufferSalida, clave, manifiestoLen)
        val entradas = deserializarManifiesto(bufferSalida.toByteArray())
        return SesionCarpeta(entradas, clave)
    }

    /**
     * Descifra todo el contenido de la sesión, en orden, invocando [crearSalida] por cada
     * entrada para obtener dónde escribirla. [entrada] debe volver a abrirse desde el
     * principio del archivo .enc (se vuelve a posicionar internamente).
     */
    fun extraerTodo(
        reabrirEntrada: () -> InputStream,
        sesion: SesionCarpeta,
        crearSalida: (EntradaManifiesto) -> OutputStream,
        progreso: ((Int, Int) -> Unit)? = null,
    ) {
        reabrirEntrada().use { entrada ->
            val inicioDatos = SALT_SIZE + 1 + 8 + longitudCifrada(manifiestoPlanoLongitud(sesion))
            saltar(entrada, inicioDatos)

            sesion.entradas.forEachIndexed { i, e ->
                crearSalida(e).use { salida -> descifrarStream(entrada, salida, sesion.clave, e.longitud) }
                progreso?.invoke(i + 1, sesion.entradas.size)
            }
        }
    }

    private fun manifiestoPlanoLongitud(sesion: SesionCarpeta): Long {
        // Se recalcula serializando de nuevo el manifiesto: es chico (solo nombres/tamaños), no los datos.
        return serializarManifiesto(sesion.entradas).size.toLong()
    }

    // ===================== Cifrado/descifrado por bloques (AES-256-GCM) =====================

    private fun cifrarStream(entrada: InputStream, salida: OutputStream, clave: ByteArray, longitudTotal: Long) {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val secretKey = SecretKeySpec(clave, "AES")
        val buffer = ByteArray(TAMANO_CHUNK)
        val nonce = ByteArray(NONCE_SIZE)
        val rng = SecureRandom()
        var restante = longitudTotal
        val chunks = numeroDeChunks(longitudTotal)

        repeat(chunks) {
            val tam = minOf(TAMANO_CHUNK.toLong(), restante).toInt()
            leerCompleto(entrada, buffer, tam)

            rng.nextBytes(nonce)
            cipher.init(Cipher.ENCRYPT_MODE, secretKey, GCMParameterSpec(TAG_SIZE * 8, nonce))
            val cifradoConTag = cipher.doFinal(buffer, 0, tam) // ciphertext + tag (16B) al final

            salida.write(nonce)
            salida.write(cifradoConTag, tam, TAG_SIZE) // tag
            salida.write(cifradoConTag, 0, tam) // ciphertext

            restante -= tam
        }
    }

    private fun descifrarStream(entrada: InputStream, salida: OutputStream, clave: ByteArray, longitudTotal: Long) {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val secretKey = SecretKeySpec(clave, "AES")
        val nonce = ByteArray(NONCE_SIZE)
        val tag = ByteArray(TAG_SIZE)
        var restante = longitudTotal
        val chunks = numeroDeChunks(longitudTotal)

        repeat(chunks) {
            val tam = minOf(TAMANO_CHUNK.toLong(), restante).toInt()
            leerCompleto(entrada, nonce)
            leerCompleto(entrada, tag)
            val cifrado = ByteArray(tam)
            leerCompleto(entrada, cifrado, tam)

            val cifradoConTag = cifrado + tag
            try {
                cipher.init(Cipher.DECRYPT_MODE, secretKey, GCMParameterSpec(TAG_SIZE * 8, nonce))
                val plano = cipher.doFinal(cifradoConTag)
                salida.write(plano)
            } catch (ex: GeneralSecurityException) {
                throw ArchivoDanadoException()
            }

            restante -= tam
        }
    }

    private fun numeroDeChunks(longitud: Long): Int =
        if (longitud == 0L) 1 else ((longitud + TAMANO_CHUNK - 1) / TAMANO_CHUNK).toInt()

    private fun longitudCifrada(longitudOriginal: Long): Long =
        longitudOriginal + numeroDeChunks(longitudOriginal).toLong() * (NONCE_SIZE + TAG_SIZE)

    // ===================== Manifiesto =====================

    private fun serializarManifiesto(entradas: List<EntradaManifiesto>): ByteArray {
        val buf = java.io.ByteArrayOutputStream()
        escribirInt(buf, entradas.size)
        for (e in entradas) {
            val pathBytes = e.relativa.toByteArray(Charsets.UTF_8)
            escribirInt(buf, pathBytes.size)
            buf.write(pathBytes)
            escribirLong(buf, e.longitud)
            escribirLong(buf, e.offset)
        }
        return buf.toByteArray()
    }

    private fun deserializarManifiesto(datos: ByteArray): List<EntradaManifiesto> {
        val stream = datos.inputStream()
        val cantidad = leerInt(stream)
        if (cantidad < 0 || cantidad > MAX_ENTRADAS) throw ArchivoDanadoException()

        val lista = ArrayList<EntradaManifiesto>(cantidad)
        repeat(cantidad) {
            val pathLen = leerInt(stream)
            if (pathLen <= 0 || pathLen > MAX_LONGITUD_RUTA) throw ArchivoDanadoException()
            val pathBytes = ByteArray(pathLen)
            leerCompleto(stream, pathBytes)
            val relativa = String(pathBytes, Charsets.UTF_8)

            val longitud = leerLong(stream)
            val offset = leerLong(stream)
            if (longitud < 0 || offset < 0) throw ArchivoDanadoException()

            lista.add(EntradaManifiesto(relativa, longitud, offset))
        }
        return lista
    }

    // ===================== Helpers de bajo nivel =====================

    private fun escribirInt(stream: OutputStream, valor: Int) {
        val bb = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(valor)
        stream.write(bb.array())
    }

    private fun escribirLong(stream: OutputStream, valor: Long) {
        val bb = ByteBuffer.allocate(8).order(ByteOrder.LITTLE_ENDIAN).putLong(valor)
        stream.write(bb.array())
    }

    private fun leerInt(stream: InputStream): Int {
        val buffer = ByteArray(4)
        leerCompleto(stream, buffer)
        return ByteBuffer.wrap(buffer).order(ByteOrder.LITTLE_ENDIAN).int
    }

    private fun leerLong(stream: InputStream): Long {
        val buffer = ByteArray(8)
        leerCompleto(stream, buffer)
        return ByteBuffer.wrap(buffer).order(ByteOrder.LITTLE_ENDIAN).long
    }

    private fun leerCompleto(stream: InputStream, buffer: ByteArray, cantidad: Int = buffer.size) {
        var leido = 0
        while (leido < cantidad) {
            val n = stream.read(buffer, leido, cantidad - leido)
            if (n <= 0) throw ArchivoDanadoException()
            leido += n
        }
    }

    private fun saltar(stream: InputStream, cantidad: Long) {
        var restante = cantidad
        while (restante > 0) {
            val saltado = stream.skip(restante)
            if (saltado <= 0) {
                if (stream.read() == -1) throw EOFException()
                restante -= 1
            } else {
                restante -= saltado
            }
        }
    }

    private fun derivarClave(password: String, salt: ByteArray): ByteArray {
        val spec = PBEKeySpec(password.toCharArray(), salt, PBKDF2_ITERATIONS, KEY_SIZE_BYTES * 8)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        return factory.generateSecret(spec).encoded
    }
}

File: linux/Encriptador/Services/CryptoService.cs
using System.IO;
using System.Security.Cryptography;
using System.Text;

namespace Encriptador.Services;

/// <summary>
/// Cifra/descifra archivos y carpetas con AES-256-GCM (autenticado, por
/// bloques de 64 KB) derivando la clave de la contraseña mediante PBKDF2
/// (600.000 iteraciones). Un archivo individual, o varios archivos sueltos
/// elegidos a la vez, o una carpeta completa (recursiva) se empaquetan en
/// un único ".enc". Las carpetas guardan un manifiesto (nombres, tamaños y
/// estructura) cifrado aparte de los datos: eso permite leer el índice y
/// abrir/extraer un archivo puntual sin descifrar el resto. Cada bloque de
/// 64 KB lleva su propio nonce y tag de autenticación, así que una
/// contraseña incorrecta o un archivo manipulado se detectan de inmediato
/// (no dependen, como en CBC, de que falle el relleno al final).
/// Por defecto se borra el original (archivo o carpeta entera) una vez
/// encriptado con éxito, salvo que se pida conservarlo; el borrado puede
/// ser "seguro" (sobrescribiendo el contenido antes de borrar).
/// </summary>
public static class CryptoService
{
    private const int SaltSize = 16;
    private const int KeySize = 32; // AES-256
    private const int Pbkdf2Iterations = 600_000;
    private const byte TipoArchivo = 1;
    private const byte TipoCarpeta = 2;
    private const int MaxEntradas = 5_000_000;
    private const int MaxLongitudRuta = 32_000;
    private const int MaxLongitudManifiesto = 200_000_000;
    private const int TamanoChunk = 64 * 1024;
    private const int NonceSize = 12;
    private const int TagSize = 16;

    public const string Extension = ".enc";

    public static bool EsArchivoEncriptado(string ruta) =>
        ruta.EndsWith(Extension, StringComparison.OrdinalIgnoreCase);

    /// <summary>Indica si un .enc contiene una carpeta (sin necesitar la contraseña: el tipo va sin cifrar).</summary>
    public static bool EsContenedorDeCarpeta(string rutaEnc)
    {
        using var fs = File.OpenRead(rutaEnc);
        if (fs.Length < SaltSize + 1)
            return false;
        fs.Seek(SaltSize, SeekOrigin.Begin);
        return fs.ReadByte() == TipoCarpeta;
    }

    // ===================== Encriptar =====================

    /// <summary>Encripta un archivo o una carpeta completa en un único ".enc".</summary>
    public static (int ArchivosIncluidos, string RutaDestino) Encriptar(
        string ruta, string password, bool conservarOriginal = false, bool borradoSeguro = false,
        IProgress<(int Actual, int Total)>? progreso = null)
    {
        if (Directory.Exists(ruta))
        {
            var raiz = ruta.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
            var destino = raiz + Extension;
            var fuente = Directory.EnumerateFiles(raiz, "*", SearchOption.AllDirectories)
                .Select(a => (RutaCompleta: a, RutaRelativa: Path.GetRelativePath(raiz, a).Replace('\\', '/')))
                .ToList();

            var cantidad = EncriptarConjunto(fuente, destino, password, progreso);

            if (!conservarOriginal)
                BorrarCarpeta(raiz, borradoSeguro);

            return (cantidad, destino);
        }

        if (File.Exists(ruta))
        {
            var destino = ruta + Extension;
            var salt = RandomNumberGenerator.GetBytes(SaltSize);
            var clave = DerivarClave(password, salt);

            using (var salida = File.Create(destino))
            {
                salida.Write(salt);
                salida.WriteByte(TipoArchivo);
                var longitud = new FileInfo(ruta).Length;
                EscribirInt64(salida, longitud);

                using var entrada = File.OpenRead(ruta);
                CifrarStream(entrada, salida, clave, longitud);
            }

            if (!conservarOriginal)
                BorrarArchivo(ruta, borradoSeguro);

            return (1, destino);
        }

        throw new FileNotFoundException(Loc.T("crypto.rutaNoExiste"), ruta);
    }

    /// <summary>
    /// Encripta varios archivos sueltos elegidos a la vez (selección múltiple) en un único ".enc"
    /// con la misma estructura que una carpeta, pero sin carpeta contenedora real.
    /// </summary>
    public static int EncriptarVarios(
        IReadOnlyList<string> rutas, string destino, string password, bool conservarOriginal = false,
        bool borradoSeguro = false, IProgress<(int Actual, int Total)>? progreso = null)
    {
        var fuente = rutas.Select(r => (RutaCompleta: r, RutaRelativa: Path.GetFileName(r))).ToList();
        var cantidad = EncriptarConjunto(fuente, destino, password, progreso);

        if (!conservarOriginal)
            foreach (var r in rutas)
                BorrarArchivo(r, borradoSeguro);

        return cantidad;
    }

    private static int EncriptarConjunto(
        List<(string RutaCompleta, string RutaRelativa)> fuente, string destino, string password,
        IProgress<(int, int)>? progreso)
    {
        var salt = RandomNumberGenerator.GetBytes(SaltSize);
        var clave = DerivarClave(password, salt);

        var entradas = new List<(string RutaCompleta, string Relativa, long Longitud, long Offset)>();
        long offsetAcumulado = 0;
        foreach (var (rutaCompleta, relativa) in fuente)
        {
            var longitud = new FileInfo(rutaCompleta).Length;
            entradas.Add((rutaCompleta, relativa, longitud, offsetAcumulado));
            offsetAcumulado += LongitudCifrada(longitud);
        }

        using var salida = File.Create(destino);
        salida.Write(salt);
        salida.WriteByte(TipoCarpeta);

        var manifiestoPlano = SerializarManifiesto(entradas);

        using (var msManifiesto = new MemoryStream(manifiestoPlano))
        {
            EscribirInt64(salida, manifiestoPlano.Length);
            CifrarStream(msManifiesto, salida, clave, manifiestoPlano.Length);
        }

        var total = entradas.Count;
        var procesados = 0;
        foreach (var e in entradas)
        {
            using (var entradaArchivo = File.OpenRead(e.RutaCompleta))
                CifrarStream(entradaArchivo, salida, clave, e.Longitud);

            procesados++;
            progreso?.Report((procesados, total));
        }

        return entradas.Count;
    }

    // ===================== Desencriptar (archivo suelto) =====================

    /// <summary>Desencripta un ".enc" de archivo individual (no carpeta) directamente a su ubicación final.</summary>
    public static (int ArchivosRestaurados, string RutaDestino) Desencriptar(string ruta, string password, bool conservarOriginal = false)
    {
        ValidarRutaEnc(ruta);

        var nombreBase = EsArchivoEncriptado(ruta) ? ruta[..^Extension.Length] : ruta + ".dec";

        using (var entrada = File.OpenRead(ruta))
        {
            var salt = new byte[SaltSize];
            LeerCompleto(entrada, salt);
            var tipo = entrada.ReadByte();
            if (tipo != TipoArchivo)
                throw new InvalidOperationException(Loc.T("crypto.tipoEsCarpeta"));

            var clave = DerivarClave(password, salt);
            var longitud = LeerInt64(entrada);
            if (longitud < 0)
                throw new CryptographicException(Loc.T("crypto.archivoDanado"));

            using var salida = File.Create(nombreBase);
            DescifrarStream(entrada, salida, clave, longitud);
        }

        if (!conservarOriginal)
            File.Delete(ruta);

        return (1, nombreBase);
    }

    // ===================== Explorar / extraer carpetas =====================

    /// <summary>
    /// Abre un .enc de carpeta (o de varios archivos): valida la contraseña y
    /// descifra únicamente el manifiesto, sin tocar el contenido de ningún
    /// archivo. Con la sesión resultante se puede listar, abrir o extraer
    /// archivos puntuales bajo demanda.
    /// </summary>
    public static SesionCarpeta AbrirCarpeta(string rutaEnc, string password)
    {
        ValidarRutaEnc(rutaEnc);

        using var entrada = File.OpenRead(rutaEnc);
        var salt = new byte[SaltSize];
        LeerCompleto(entrada, salt);
        var tipo = entrada.ReadByte();
        if (tipo != TipoCarpeta)
            throw new InvalidOperationException(Loc.T("crypto.tipoEsArchivo"));

        var clave = DerivarClave(password, salt);

        var manifiestoLen = LeerInt64(entrada);
        if (manifiestoLen < 0 || manifiestoLen > MaxLongitudManifiesto)
            throw new CryptographicException(Loc.T("crypto.archivoDanado"));

        byte[] manifiestoPlano;
        using (var msSalida = new MemoryStream())
        {
            DescifrarStream(entrada, msSalida, clave, manifiestoLen);
            manifiestoPlano = msSalida.ToArray();
        }

        var inicioDatos = entrada.Position;
        var entradas = DeserializarManifiesto(manifiestoPlano);

        return new SesionCarpeta(rutaEnc, clave, inicioDatos, entradas);
    }

    /// <summary>
    /// Sesión sobre un .enc de carpeta ya validado: permite listar, abrir o
    /// extraer archivos puntuales sin volver a pedir la contraseña.
    /// </summary>
    public sealed class SesionCarpeta
    {
        private readonly string _rutaEnc;
        private readonly byte[] _clave;
        private readonly long _inicioDatos;
        private readonly List<EntradaManifiesto> _entradas;

        internal SesionCarpeta(string rutaEnc, byte[] clave, long inicioDatos, List<EntradaManifiesto> entradas)
        {
            _rutaEnc = rutaEnc;
            _clave = clave;
            _inicioDatos = inicioDatos;
            _entradas = entradas;
        }

        /// <summary>Rutas relativas (con '/') y tamaño original de cada archivo incluido.</summary>
        public IReadOnlyList<(string RutaRelativa, long Longitud)> Archivos =>
            _entradas.Select(e => (e.Relativa, e.Longitud)).ToList();

        /// <summary>Descifra un único archivo (por su ruta relativa) al destino indicado.</summary>
        public void ExtraerArchivo(string rutaRelativa, string destino)
        {
            var entrada = _entradas.FirstOrDefault(e => e.Relativa == rutaRelativa)
                ?? throw new FileNotFoundException(Loc.T("crypto.archivoNoManifiesto"), rutaRelativa);

            var carpetaContenedora = Path.GetDirectoryName(destino);
            if (!string.IsNullOrEmpty(carpetaContenedora))
                Directory.CreateDirectory(carpetaContenedora);

            using var origen = File.OpenRead(_rutaEnc);
            origen.Seek(_inicioDatos + entrada.Offset, SeekOrigin.Begin);

            using var salida = File.Create(destino);
            DescifrarStream(origen, salida, _clave, entrada.Longitud);
        }

        /// <summary>Descifra varios archivos (rutas relativas) dentro de <paramref name="carpetaDestino"/>, preservando su estructura.</summary>
        public void ExtraerVarios(IEnumerable<string> rutasRelativas, string carpetaDestino, IProgress<(int Actual, int Total)>? progreso = null)
        {
            var lista = rutasRelativas.ToList();
            var procesados = 0;
            foreach (var relativa in lista)
            {
                var destino = Path.Combine(carpetaDestino, relativa.Replace('/', Path.DirectorySeparatorChar));
                ExtraerArchivo(relativa, destino);
                procesados++;
                progreso?.Report((procesados, lista.Count));
            }
        }
    }

    internal sealed record EntradaManifiesto(string Relativa, long Longitud, long Offset);

    // ===================== Cifrado/descifrado por bloques (AES-256-GCM) =====================

    private static void CifrarStream(Stream entrada, Stream salida, byte[] clave, long longitudTotal)
    {
        using var aesGcm = new AesGcm(clave, TagSize);
        var buffer = new byte[TamanoChunk];
        var nonce = new byte[NonceSize];
        var tag = new byte[TagSize];
        var restante = longitudTotal;

        var chunks = NumeroDeChunks(longitudTotal);
        for (var i = 0; i < chunks; i++)
        {
            var tam = (int)Math.Min(TamanoChunk, restante);
            LeerCompleto(entrada, buffer.AsSpan(0, tam));

            RandomNumberGenerator.Fill(nonce);
            var cifrado = new byte[tam];
            aesGcm.Encrypt(nonce, buffer.AsSpan(0, tam), cifrado, tag);

            salida.Write(nonce);
            salida.Write(tag);
            salida.Write(cifrado);

            restante -= tam;
        }
    }

    private static void DescifrarStream(Stream entrada, Stream salida, byte[] clave, long longitudTotal)
    {
        using var aesGcm = new AesGcm(clave, TagSize);
        var nonce = new byte[NonceSize];
        var tag = new byte[TagSize];
        var restante = longitudTotal;

        var chunks = NumeroDeChunks(longitudTotal);
        for (var i = 0; i < chunks; i++)
        {
            var tam = (int)Math.Min(TamanoChunk, restante);
            LeerCompleto(entrada, nonce);
            LeerCompleto(entrada, tag);
            var cifrado = new byte[tam];
            LeerCompleto(entrada, cifrado);

            var plano = new byte[tam];
            try
            {
                aesGcm.Decrypt(nonce, cifrado, tag, plano);
            }
            catch (CryptographicException)
            {
                throw new CryptographicException(Loc.T("crypto.archivoDanado"));
            }

            salida.Write(plano);
            restante -= tam;
        }
    }

    private static int NumeroDeChunks(long longitud) =>
        longitud == 0 ? 1 : (int)((longitud + TamanoChunk - 1) / TamanoChunk);

    private static long LongitudCifrada(long longitudOriginal) =>
        longitudOriginal + (long)NumeroDeChunks(longitudOriginal) * (NonceSize + TagSize);

    // ===================== Borrado (normal y seguro) =====================

    /// <summary>Borra una carpeta temporal (p. ej. de vista previa) sobrescribiendo su contenido antes. Falla en silencio si algún archivo sigue abierto en otra app.</summary>
    public static void LimpiarCarpetaTemporalSegura(string carpeta)
    {
        try
        {
            if (Directory.Exists(carpeta))
                BorrarCarpeta(carpeta, seguro: true);
        }
        catch
        {
            // Best-effort: puede seguir abierto en otra app.
        }
    }

    private static void BorrarArchivo(string ruta, bool seguro)
    {
        if (seguro)
            SobrescribirConAleatorio(ruta);
        File.Delete(ruta);
    }

    private static void BorrarCarpeta(string carpeta, bool seguro)
    {
        if (seguro)
            foreach (var archivo in Directory.EnumerateFiles(carpeta, "*", SearchOption.AllDirectories))
                SobrescribirConAleatorio(archivo);

        Directory.Delete(carpeta, recursive: true);
    }

    private static void SobrescribirConAleatorio(string ruta)
    {
        var longitud = new FileInfo(ruta).Length;
        using var fs = new FileStream(ruta, FileMode.Open, FileAccess.Write, FileShare.None);
        var buffer = new byte[81920];
        var restante = longitud;
        while (restante > 0)
        {
            var n = (int)Math.Min(buffer.Length, restante);
            RandomNumberGenerator.Fill(buffer.AsSpan(0, n));
            fs.Write(buffer, 0, n);
            restante -= n;
        }
        fs.Flush(flushToDisk: true);
    }

    // ===================== Helpers internos =====================

    private static void ValidarRutaEnc(string ruta)
    {
        if (Directory.Exists(ruta))
            throw new InvalidOperationException(Loc.T("crypto.seleccionarArchivoEnc"));

        if (!File.Exists(ruta))
            throw new FileNotFoundException(Loc.T("crypto.rutaNoExiste"), ruta);
    }

    private static byte[] SerializarManifiesto(List<(string RutaCompleta, string Relativa, long Longitud, long Offset)> entradas)
    {
        using var ms = new MemoryStream();
        EscribirInt32(ms, entradas.Count);
        foreach (var e in entradas)
        {
            var pathBytes = Encoding.UTF8.GetBytes(e.Relativa);
            EscribirInt32(ms, pathBytes.Length);
            ms.Write(pathBytes);
            EscribirInt64(ms, e.Longitud);
            EscribirInt64(ms, e.Offset);
        }
        return ms.ToArray();
    }

    private static List<EntradaManifiesto> DeserializarManifiesto(byte[] datos)
    {
        using var ms = new MemoryStream(datos);
        var cantidad = LeerInt32(ms);
        if (cantidad < 0 || cantidad > MaxEntradas)
            throw new CryptographicException(Loc.T("crypto.archivoDanado"));

        var lista = new List<EntradaManifiesto>(cantidad);
        for (var i = 0; i < cantidad; i++)
        {
            var pathLen = LeerInt32(ms);
            if (pathLen <= 0 || pathLen > MaxLongitudRuta)
                throw new CryptographicException(Loc.T("crypto.archivoDanado"));

            var pathBytes = new byte[pathLen];
            LeerCompleto(ms, pathBytes);
            var relativa = Encoding.UTF8.GetString(pathBytes);

            var longitud = LeerInt64(ms);
            var offset = LeerInt64(ms);
            if (longitud < 0 || offset < 0)
                throw new CryptographicException(Loc.T("crypto.archivoDanado"));

            lista.Add(new EntradaManifiesto(relativa, longitud, offset));
        }
        return lista;
    }

    private static void EscribirInt32(Stream stream, int valor) => stream.Write(BitConverter.GetBytes(valor));

    private static void EscribirInt64(Stream stream, long valor) => stream.Write(BitConverter.GetBytes(valor));

    private static int LeerInt32(Stream stream)
    {
        Span<byte> buffer = stackalloc byte[4];
        LeerCompleto(stream, buffer);
        return BitConverter.ToInt32(buffer);
    }

    private static long LeerInt64(Stream stream)
    {
        Span<byte> buffer = stackalloc byte[8];
        LeerCompleto(stream, buffer);
        return BitConverter.ToInt64(buffer);
    }

    private static void LeerCompleto(Stream stream, Span<byte> buffer)
    {
        var leido = 0;
        while (leido < buffer.Length)
        {
            var n = stream.Read(buffer[leido..]);
            if (n == 0)
                throw new CryptographicException(Loc.T("crypto.archivoDanado"));
            leido += n;
        }
    }

    private static byte[] DerivarClave(string password, byte[] salt)
    {
        using var pbkdf2 = new Rfc2898DeriveBytes(password, salt, Pbkdf2Iterations, HashAlgorithmName.SHA256);
        return pbkdf2.GetBytes(KeySize);
    }
}
